pkg/client/unversioned/portforward/portforward.go

```diff
@@ -250,67 +254,79 @@ func (pf *PortForwarder) waitForConnection(listener net.Listener, port Forwarded
+		switch {
+		case err != nil:
+			errorChan <- fmt.Errorf("error reading from error stream for port %d -> %d: %v", port.Local, port.Remote, err)
+		case len(message) > 0:
+			errorChan <- fmt.Errorf("an error occurred forwarding %d -> %d: %v", port.Local, port.Remote, string(message))
+		default:
+			errorChan <- nil
 		}
+		close(errorChan)
 	}()
 
 	// create data stream
 	headers.Set(api.StreamType, api.StreamTypeData)
 	dataStream, err := pf.streamConn.CreateStream(headers)
 	if err != nil {
-		glog.Errorf("Error creating forwarding stream for port %d -> %d: %v", port.Local, port.Remote, err)
+		util.HandleError(fmt.Errorf("error creating forwarding stream for port %d -> %d: %v", port.Local, port.Remote, err))
 		return
 	}
-	// Send a Reset when this function exits to completely tear down the stream here
-	// and in the remote server.
-	defer dataStream.Reset()
+
+	doneChan := make(chan struct{}, 2)
 
 	go func() {
-		// Copy from the remote side to the local port. We won't get an EOF from
-		// the server as it has no way of knowing when to close the stream.  We'll
-		// take care of closing both ends of the stream with the call to
-		// stream.Reset() when this function exits.
-		if _, err := io.Copy(conn, dataStream); err != nil && err != io.EOF && !strings.Contains(err.Error(), "use of closed network connection") {
-			glog.Errorf("Error copying from remote stream to local connection: %v", err)
+		// Copy from the remote side to the local port.
+		if _, err := io.Copy(conn, dataStream); err != nil && !strings.Contains(err.Error(), "use of closed network connection") {
+			util.HandleError(fmt.Errorf("error copying from remote stream to local connection: %v", err))
 		}
+
+		// inform the select below that the remote copy is done
 		doneChan <- struct{}{}
 	}()
 
 	go func() {
-		// Copy from the local port to the remote side. Here we will be able to know
-		// when the Copy gets an EOF from conn, as that will happen as soon as conn is
-		// closed (i.e. client disconnected).
-		if _, err := io.Copy(dataStream, conn); err != nil && err != io.EOF && !strings.Contains(err.Error(), "use of closed network connection") {
-			glog.Errorf("Error copying from local connection to remote stream: %v", err)
+		// inform server we're not sending any more data after copy unblocks
+		defer dataStream.Close()
+
+		// Copy from the local port to the remote side.
+		if _, err := io.Copy(dataStream, conn); err != nil && !strings.Contains(err.Error(), "use of closed network connection") {
+			util.HandleError(fmt.Errorf("error copying from local connection to remote stream: %v", err))
+			// break out of the select below without waiting for the other copy to finish
+			doneChan <- struct{}{}
 		}
-		doneChan <- struct{}{}
 	}()
 
-	select {
-	case err := <-errorChan:
-		glog.Error(err)
-	case <-doneChan:
+	// wait for either a local->remote error or for copying from remote->local to finish
+	<-doneChan
```

Comment:
It's a little more verbose, but I think the semantics of what you want might come through better if you use separate channels and select on them. Like

```go
localError := make(chan struct{})
remoteDone := make(chan struct{})
...
go func() {
	...
	close(remoteDone)
}()
...
go func() {
	if _, err := ...; err != nil {
		close(localError)
	}
}()

select {
case <-remoteDone:
case <-localError:
} 

```